### DISCUSSION TITLE
Intercept build commands using ptrace

### DISCUSSION THREAD

**Discussion by i-ky** [rizsotto/Bear#522] [Show and tell]
Some time ago I was in a search for a way to generate `compile_commands.json` for a project with a very tricky build system that did not respect `CC`/`CXX`, used hardcoded compiler paths and statically compiled binaries here and there. Unfortunately, `bear` is not able to intercept commands in this case. Eventually I came across [`compile-db-gen`](https://github.com/sunlin7/compile-db-gen) which uses `strace` and then developed [my own solution](https://github.com/i-ky/basset) based on `ptrace`. I wonder if you would be interested in adding such command interception mechanism to `bear`. If yes, I will be glad to help.

**Comment by rizsotto** [comment]
Hey @i-ky , this is a good proposal... Bear is already using 2 interception mode (the compiler wrappers and the library preload), adding another one would increase the chance to cover more builds. And Bear is already split the work of recognizing compiler calls and intercepting process executions.

Was looking at your project, it looks very simple. I would be glad to review PRs on this work, if you have time. (I won't be able to work on this in the near future, because I am traveling at the moment.)

**Reply by i-ky** [reply]
> And Bear is already split the work of recognizing compiler calls and intercepting process executions.

I am aware of that, but unfortunately the [output format of `bear intercept`](https://github.com/rizsotto/Bear/blob/master/source/intercept/man/bear-intercept.1.md#output-file) isn't described in detail:
> The output file has [JSON lines](https://jsonlines.org/) format, where each line terminated with \n line separator and each line is a JSON object.
>
> The JSON objects are process execution events: process start, process got signal, process terminated. (For the schema of these events, please consult with the source code of this project.)

I guess that there are reasons to keep exact schema as implementation detail of `bear`, but on the other hand properly documenting it would make it easier to create external drop-in replacements of `bear intercept`.

> I would be glad to review PRs on this work, if you have time.

I will put it on my TODO list, but no promises on the time frame. I will be glad if you share your ideas on the best way to integrate `basset` into `bear`.

**Reply by rizsotto** [reply]
> > And Bear is already split the work of recognizing compiler calls and intercepting process executions.
> 
> I am aware of that, but unfortunately the [output format of `bear intercept`](https://github.com/rizsotto/Bear/blob/master/source/intercept/man/bear-intercept.1.md#output-file) isn't described in detail:
> 
> > The output file has [JSON lines](https://jsonlines.org/) format, where each line terminated with \n line separator and each line is a JSON object.
> > The JSON objects are process execution events: process start, process got signal, process terminated. (For the schema of these events, please consult with the source code of this project.)
> 
> I guess that there are reasons to keep exact schema as implementation detail of `bear`, but on the other hand properly documenting it would make it easier to create external drop-in replacements of `bear intercept`.

Well, I did not see big demand for this file format to be open. :smile:

As the quoted document says, it's an list of JSON objects separated by newline character. The JSON objects are the gRPC events from this file: https://github.com/rizsotto/Bear/blob/master/source/intercept/proto/intercept.proto

To have a better understanding how the file looks like, just run the `bear intercept` command and inspect the output.

> > I would be glad to review PRs on this work, if you have time.
> 
> I will put it on my TODO list, but no promises on the time frame. I will be glad if you share your ideas on the best way to integrate `basset` into `bear`.

That sounds good to me.

I have a few questions that needs to be clarified before we get into design.... One of them is how the working directory captured in the ptrace case? What happens if the process which executes the compiler changing the working directory?

**Reply by i-ky** [reply]
> I have a few questions that needs to be clarified before we get into design.... One of them is how the working directory captured in the ptrace case? What happens if the process which executes the compiler changing the working directory?

Here is how `basset` works. After parsing command line arguments it `fork()`'s. Parent process [sets `ptrace` options on the child process](https://github.com/i-ky/basset/blob/a20fa63067d6d506b59c853830997dfc7284547a/src/basset.cpp#L300-L302) and then signals the child, that it can continue. Child process waits for a signal from parent and then `execvp()`'s the build command. (I say "signal", but there is no signal in UNIX sense, parent actually sends a single byte to the child via a pipe. Since `ptrace` (ab)uses `SIGTRAP` to inform tracer about its own events, so less signals flying around makes life a bit easier.)

Specified `ptrace` options (`PTRACE_O_TRACECLONE`, `PTRACE_O_TRACEFORK`, `PTRACE_O_TRACEVFORK`, `PTRACE_O_TRACEEXEC`) ensure that processes being `ptrace`'d are stopped when they do `clone()`, `fork()`, `vfork()` and `execve()`. The latter kind of stop (`PTRACE_EVENT_EXEC`) is of the most interest for `compile_commands.json` generation, because it allows tracer process to [read executable path from `/proc/[pid]/exe`](https://github.com/i-ky/basset/blob/a20fa63067d6d506b59c853830997dfc7284547a/src/basset.cpp#L339-L340), [working directory from `/proc/[pid]/cwd`](https://github.com/i-ky/basset/blob/a20fa63067d6d506b59c853830997dfc7284547a/src/basset.cpp#L354-L355) and [command line from `/proc/[pid]/cmdline`](https://github.com/i-ky/basset/blob/a20fa63067d6d506b59c853830997dfc7284547a/src/basset.cpp#L362-L367). Since the command being executed has not yet started execution, there is no chance it could `chdir()` or alter its command line arguments.

**Reply by rizsotto** [reply]
Thanks for the explanation @i-ky . As a follow up question, is this a linux specific solution only? Or can work with other OS too? (The [wiki page](https://en.wikipedia.org/wiki/Ptrace) mentions that ptrace itself is available on other platforms too. But wondering if the `/proc/[pid]/*` files are standard locations across OSes?)

**Reply by i-ky** [reply]
> As a follow up question, is this a linux specific solution only?

Yes, `/proc/[pid]/{exe,cwd,cmdline}` seem to be Linux-only.

**Reply by i-ky** [reply]
Pure-`ptrace` solution would support more OSes, but it is way more complicated, because it would need to track `chdir()` syscalls and this is not a trivial task since `ptrace` only provides syscall numbers which vary depending on the architecture. Building on top of `strace` would be an option, but https://github.com/strace/strace/issues/70 or https://github.com/strace/strace/issues/169 needs to be implemented first.

**Reply by rizsotto** [reply]
I was looking into run `strace` and parse its output, but failed to get the full argument list of the function calls.

**Reply by i-ky** [reply]
`compile-db-gen` uses `-s`/`--string-limit` option and passes the result of `getconf ARG_MAX` as a value, see [here](https://github.com/sunlin7/compile-db-gen/blob/75f0a3825706ee2a0769d85c33584ddec06e6f10/compile-db-gen.py#L286-L292). This tactic works, but `compile-db-gen` has a couple of other problems. Not to mention, that parsing `strace` output with regular expressions is neither efficient nor robust.

**Reply by fpoirotte** [reply]
> `ptrace` only provides syscall numbers which vary depending on the architecture.

On Linux, `/usr/include/asm/unistd*.h` (provided by kernel-headers on Fedora)  can be used to retrieve syscall numbers.

However, this may prove difficult to use (though I think it is pretty rare, the number associated with a syscall can theoretically vary depending on the kernel's version -- also, new syscalls may be added over time). This would probably require parsing the file at runtime to build a mapping of available syscalls and their number.

**Reply by i-ky** [reply]
@fpoirotte, difficulties don't end with syscall numbers. Imagine you identified a syscall of interest (`chdir()` or `exec()`), you than need to peek into tracee's memory to figure out the arguments of the syscall (the directory it changes into or executable it tries to execute). `ptrace` allows you to do it, but only one word at a time, which is quite inefficient. `strace` has implemented various platform-dependent optimizations and only falls back to word-by-word reads in case nothing better is available. It would be nice to have `strace` as a library, otherwise reinventing the wheel would be quite a challenge.
